### PR TITLE
fix: Correct the AMI selector label requirements for NVIDIA and Neuron variants

### PR DESCRIPTION
**Pull request by bryantbiggs**
Fixes #N/A 

**Description**

- Correct the AMI selector label requirements for NVIDIA and Neuron variants
  - Instances backed with NVIDIA GPUs should have the label `karpenter.sh/instance-gpu-count`
  - Instances backed with Neuron devices should have the label `karpenter.sh/instance-accelerator-count`

**How was this change tested?**

- `make test`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jmdeal — APPROVED** [review]
LGTM 🚀 

Looks like this slipped through because our currently supported AMIs don't differentiate between Nvidia and Neuron, so this swap doesn't change the end result. I'd be happy to merge and take a test for that case as a follow-up (separate Nvidia and Neuron AMIs, same AMI family), but also happy to take one here if you wanted to add one.

**Comment by bryantbiggs** [comment]
I'm dreadfully awful with go 😅 - give me 24 hours and let me see if I can add in tests for this

**Comment by bryantbiggs** [comment]
@jmdeal I think this one is outside of my wheelhouse - I tried a naive approach to check that the `scheduling.Requirements` included the appropriate label, but wasn't able to make it work. I'm afraid this one is outside of my wheelhouse 😅

**Comment by jmdeal** [comment]
Sorry for the delay, made a couple additional changes and this should be good to go from my perspective. Since I've touched the PR though I shouldn't be the one to approve, I'm going to get review from someone else on the team. We should be able to merge this shortly

File: pkg/controllers/nodeclass/status/ami_test.go
Comment (njtran) [inline pkg/controllers/nodeclass/status/ami_test.go:509]: nit
```suggestion
	It("should resolve amiSelector AMIs and requirements into status when all SSM parameters don't resolve", func() {
```

File: pkg/controllers/nodeclass/status/ami_test.go
Comment (njtran) [inline pkg/controllers/nodeclass/status/ami_test.go:195]: do we need a Accelerator NotExists in Accelerator here?

File: pkg/controllers/nodeclass/status/ami_test.go
Comment (njtran) [inline pkg/controllers/nodeclass/status/ami_test.go:210]: same with do we need a NotInGPU here too?

File: pkg/controllers/nodeclass/status/ami_test.go
Comment (njtran) [inline pkg/controllers/nodeclass/status/ami_test.go:249]: Do you really need to redefine it? Can't you just use the AMIs that you defined up above in a variable and reference it? Saves on code lines.

File: pkg/controllers/nodeclass/status/ami_test.go
Comment (njtran) [inline pkg/controllers/nodeclass/status/ami_test.go:404]: so this was only an issue with al2023, yes?

Comment (bryantbiggs) [reply pkg/controllers/nodeclass/status/ami_test.go:404]: looks like this note has a typo `Bottlerocket` -> `AL2`

Comment (jmdeal) [reply pkg/controllers/nodeclass/status/ami_test.go:404]: The only accelerated AMI vended by Bottlerocket is the nvidia variant so this wasn't a typo. We've historically selected the nvidia AMI for both nvidia and inferentium / trainium instances but looking at this [issue](https://github.com/bottlerocket-os/bottlerocket/issues/1995) I'm wondering if that was incorrect.

But yes, the only AMI family we currently support that will have separate AMIs for gpu and accelerator workloads is AL2023.

Comment (bryantbiggs) [reply pkg/controllers/nodeclass/status/ami_test.go:404]: > We've historically selected the nvidia AMI for both nvidia and inferentium / trainium instances but looking at this https://github.com/bottlerocket-os/bottlerocket/issues/1995 I'm wondering if that was incorrect.

Thats correct - only the EKS AL2 GPU AMI can be used on instances with NVIDIA GPUs and AWS Neuron devices. The Bottlerocket NVIDIA variant will only work on NVIDIA GPUs; it won't work on AWS Neuron devices (though that variant is in the works)

File: pkg/controllers/nodeclass/status/ami_test.go
Comment (jmdeal) [reply pkg/controllers/nodeclass/status/ami_test.go:195]: It wouldn't be incorrect, but I don't think it's necessary. AFAIK the two labels are mutually exclusive; the existence of one implies the absence of the other.

File: pkg/controllers/nodeclass/status/ami_test.go
Comment (jmdeal) [reply pkg/controllers/nodeclass/status/ami_test.go:249]: When you say up above, are you referring to the expected AMIs in the status conditions from AL2023? In that particular example they're different since AL2023 has different AMIs for neuron and nvidia. Even if they were the same, I don't think we want to couple them in the test since it's really just happenstance. I.e. if the AMIs available for one family changed, that doesn't mean the AMIs for another would change in kind.

File: pkg/controllers/nodeclass/status/ami_test.go
Comment (bryantbiggs) [reply pkg/controllers/nodeclass/status/ami_test.go:377]: > do you know if this AMI is strictly incompatible? 

Yes - you can only use the Bottlerocket NVIDIA variant on instances with NVIDIA GPUs and vice versa with Neuron devices *if you want to use those GPUs/accelerators. And no, nothing can fix this in user data - the NVIDIA driver and Neuron driver are baked into the OS which is immutable once created

> Either way, I think we can take that as a follow-up rather than changing that here.

Yes - I will defer to you all in terms of how to best tackle this

Comment (jmdeal) [reply pkg/controllers/nodeclass/status/ami_test.go:377]: Discussed with the team today, we'll plan on making this change at v1.1.0. While it shouldn't affect any users, and it is addressing a bug, it's still a breaking change so I'd like to all it out as such. Since we're going to be backporting this PR to 1.0, we'll track that change in a separate PR.

**Review by njtran — APPROVED** [review]

**Pull request merged**
